invitations: Default to empty channel list when inviting guest users.

saubhagya-patel opened the pull request:
This PR resets the channels list to be empty (with the default channels checkbox unchecked) when the user selects "Guests" from the role dropdown while inviting users. It includes the following changes:

- When inviting users for guest roles, an empty channel list is displayed instead of the default channel list, and the default channels checkbox is unchecked.
- When inviting users for non-guest roles, the default channels checkbox remains selected (maintaining the old behavior).
- When switching between guest and non-guest roles, the previous state of included channels (if any specified explicitly by the user) for the guest role is restored while inviting a user as a guest.


Fixes: #31988.

**Screenshots and screen captures:**
**Before Changes**
<details>
<summary>1. Inviting Users as Guest</summary>

![guest before](https://github.com/user-attachments/assets/e9dda498-2e31-4446-81d1-3b1b332cb35b)


</details>

<details>
<summary>2. Inviting Users for Non-Guest Roles</summary>

![non guest before](https://github.com/user-attachments/assets/1274f8d0-eab2-4b2c-8dcf-ef8062972dd1)

</details>

<details>
<summary>3. Switching between Guest and Non-Guest Roles</summary>

![switch before](https://github.com/user-attachments/assets/2ae6a259-fdd3-4109-a77c-b1c6fb9f192e)

</details>


**After Changes**
<details>
<summary>1. Inviting Users as Guest</summary>

![guest after](https://github.com/user-attachments/assets/5d886260-5523-481c-863a-c71c655a8c95)

</details>

<details>
<summary>2. Inviting Users for Non-Guest Roles (no changes)</summary>

![non guest after](https://github.com/user-attachments/assets/94ee68a7-f2ca-4bef-9e00-1696aea49534)


</details>

<details>
<summary>3. Switching between Guest and Non-Guest Roles</summary>

Please note that if the channels named _announce_ and _Denmark_ are selected for the "Guests" role and then the role is changed to "Members" when switching back to the guest role, the previously selected channels are restored in the channel list.

![switch after](https://github.com/user-attachments/assets/5004198b-6518-402c-90ea-fe01962fe9fd)

</details>



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

saubhagya-patel commented:
@alya as mentioned in [this comment](https://github.com/zulip/zulip/pull/32153#issuecomment-2446000673), I have collaborated with @userAdityaa and mentioned him as a co-author in my commits. This PR is ready for review. Thank you.

userAdityaa commented:
Hey @alya, Ready for review!

prakhar1144 commented:
Thanks @saubhagya-patel @userAdityaa !

@sahil839, would you be up for reviewing this one?

sahil839 commented:
I am not sure about the expected behavior. The issue mentions that when switching between roles we should go back to previous state of what channels are included. So, does it mean that the list should become empty everytime I switch the role dropdown to guest from any other role and then when I change the role from guest to some other role, the list of streams should be what it was before the role was changed to guest?

saubhagya-patel commented:
The current behavior when inviting users is that the default channels are selected (the "default channels" checkbox is selected) for both Guest and Non-Guest roles.

Based on the changes discussed in [the issue](https://github.com/zulip/zulip/issues/31988) and the [CZO thread](https://chat.zulip.org/#narrow/channel/101-design/topic/uncheck.20.22default.20channels.22.20for.20guests.3F/near/1961543), the following changes have been addressed:

- When inviting users for non-guest roles, the default channels are selected (the "default channels" checkbox is selected) - _no changes to the current behavior_.
- When inviting users for guest roles, the channel list should be empty, with the "default channels" checkbox unselected, if the user has not explicitly specified any channels.
- If a user specifies one or more channels explicitly (e.g., _#announce_ and _#Denmark_) when inviting for a guest role, then:
  - When the user switches to a non-guest role, the "default channels" checkbox should be selected, and the channel list should include the default channels.
  - When switching back to the guest role, the channel list should restore the channels that the user had previously specified (e.g., _#announce_ and _#Denmark_).

Kindly check the visual changes mentioned in [this comment](https://github.com/zulip/zulip/pull/32167#issue-2622083516).

Hence,
> So, does it mean that the list should become empty everytime I switch the role dropdown to guest from any other role

The list should restore the channels specified for Guest roles, if any. So, if no channels are explicitly specified while inviting for a guest role the channel list would be empty while switching back to a guest role from a non-guest role.

> and then when I change the role from guest to some other role, the list of streams should be what it was before the role was changed to guest?

When we switch from a guest role to a non-guest role the list follows current behavior, i.e. the default channels are selected (the "default channels" checkbox is selected).

@sahil839 I hope I was able to highlight the expected behavior and the changes made. Thank you.

sahil839 reviewed (commented):
Posted one comment.

Regarding the behavior, do we also want to retain the streams selected for non-guest role while changing the role? Like I selected a couple of streams when role was selected as "Members". Then I change the role to "Guest" and then again change the role to "Members", default streams are selected again. Do we want to have the same couple of streams that I selected initially. Looks good otherwise.

saubhagya-patel commented:
I have updated the PR as mentioned in [this](https://github.com/zulip/zulip/pull/32167#discussion_r1846205732) comment.
- Change made: Renamed the function `update_guest_visible_users_count()` to `update_guest_visible_users_count_and_stream_ids()`.

As @sahil839 mentioned in [this](https://github.com/zulip/zulip/pull/32167#pullrequestreview-2442044530) comment:
> Regarding the behavior, do we also want to retain the streams selected for non-guest role while changing the role? Like I selected a couple of streams when role was selected as "Members". Then I change the role to "Guest" and then again change the role to "Members", default streams are selected again. Do we want to have the same couple of streams that I selected initially.

I don't think we want to retain the streams selected for non-guest roles when changing roles for invited users. This is based on the [issue](https://github.com/zulip/zulip/issues/31988) and the [CZO thread](https://chat.zulip.org/#narrow/channel/101-design/topic/uncheck.20.22default.20channels.22.20for.20guests.3F/near/1961543), both of which suggest modifying the channel list for guest roles.
Please let me know if any adjustments are needed. Thank you.

sahil839 commented:
Ok, code changes look good as per the current behavior.

timabbott commented:
Merged, thanks @saubhagya-patel!
